src/runtime/execution.rs

```diff
@@ -307,18 +308,24 @@ impl ExecutionState {
             (std::mem::replace(&mut state.tasks, SmallVec::new()), state.current_task)
         });
 
+        // NOTE: This needs to be set before tasks are dropped
+        Self::with(|state| state.has_cleaned_up = true);
```

Comment:
I think rather than abusing this field we should just factor out a method like
```
pub(crate) fn is_finished() -> bool {
    state.current_task == ScheduledTask::Stopped || state.current_task == ScheduledTask::Finished
}
```